packages/playwright-ct-angular/index.d.ts

```diff
@@ -57,6 +58,10 @@ interface MountResult<Component> extends Locator {
 }
 
 export interface ComponentFixtures {
+  mount<HooksConfig extends JsonObject, Component = unknown>(
+    template: string,
```

Comment:
Template and imports go along.
If fixing imports support comes without additional effort to supporting providers, then we can just ship everything.
Otherwise, we can ship this later.

IMO, using templates is critical for the adoption of playwright CT.
It leverages testing techniques that are meaningful and reduce the boilerplate of test container components and outputs testing

Reply:
Fixed

Reply:
Could you help me understand the advantages and disadvantages of the template API compared to the current/object API a bit more? 

> It leverages testing techniques that are meaningful

I think that the template API aligns a bit better with how the actual components are used. Are there other techniques that are better/more meaningful compared to the Object API?

> reduce the boilerplate of test container components and outputs testing

This appears identical in my view. Can you provide an example where this scenario applies?

I wanted to dive deeper into it, but the biggest downside i see ATM is that there is no type checking with the template API, correct? Unless there is an LSP plugin available for the code editor/IDE

Reply:
> Could you help me understand the advantages and disadvantages of the template API compared to the current/object API a bit more?

The main advantage is that it avoids creating wrapper components and the associated boilerplate while keeping the tests closer to the usage as you mentioned, typically when testing directives or pipes but also components in general.

Typical example is testing a directive :
```ts
mount("<p highlightTypos>{{value}}<p>", {
  imports: [HighlightTyposDirective],
  componentProperties: {value: 'something'}
});
```
vs.
```ts
@Component({
  standalone: true,
  imports: [HighlightTyposDirective],
  template: `<p highlightTypos>{{value}}<p>`
})
class WrapperComponent {
  value = input.required<string>();
}

mount(WrapperComponent, {
  componentProperties: {value: 'something'}
});
```

The main disadvantages are that:
- it requires JIT
- it is not type-safe

> Are there other techniques that are better/more meaningful compared to the Object API?

You mean something like JSX?

> This appears identical in my view. Can you provide an example where this scenario applies?

The output testing could be tested differently so instead of passing a callback and polling like this (maybe because there is a debounce delay or something):
```ts
let submitted = false;

const component = await mount(ButtonComponent, {
  on: {
    submit() { submitted = true; },
  },
});

await component.getByRole('button').click();
await expect.poll(() => submitted).toBe(true);
```

one could do this:
```ts
mount(`
  <app-button (submit)="submitted = true"/>
  <p>{{submitted}}</p>
`, {imports: [ButtonComponent]);

await component.getByRole('button').click();
await component.getByRole('paragraph').toHaveText('true')
```

> I wanted to dive deeper into it, but the biggest downside i see ATM is that there is no type checking with the template API, correct? Unless there is an LSP plugin available for the code editor/IDE

Indeed, it is not type safe but WebStorm's Angular plugin does a great job not only with syntax highlight but also refactoring (renaming an input for example will update the tests).

Reply:
Hey, thanks for the explanations. Directives and pipes are a nice touch. However, I think that needs to be looked at separately because it's not supported in other frameworks like Vue either.

If it boils down to a type-safe API versus an API that mirrors the actual code, then I would rather stick with the type-safe API for now.

The newly introduced API’s seems to be the last thing. Maybe we should just merge without them like we discussed before so people can start using it already? They can still be added later on if the Playwright team agrees with it

Reply:
Hey, I believe that type safety represents a much smaller disadvantage compared to the necessity of creating a wrapper component every time we want to test a directive or a pipe. On a large codebase, this can lead to a significant amount of boilerplate code, just to encapsulate the element we want to test.

Reply:
Hi @edbzn, not suggesting that creating a wrapper component every time should be the final solution for testing directives and pipes directly. I agree that it can be quite cumbersome to create wrappers.

Do you see pipes/directives being tested directly that often? You are probably aware, but it's worth noting that directives and pipes requiring registration can still be defined in the `hooksConfig`;

```ts
@Component({ template: `<p highlightTypos>{{value}}<p>` })
class WrapperComponent { value = input.required<string>(); }

mount(WrapperComponent, {
  props: {value: 'something'}
  hooksConfig: {
     imports: [HighlightTyposDirective]
   }
});
```

Reply:
It depends on your testing strategy, but in general, it's recommended to test pipes and directives. Moreover, as @yjaaidi mentioned, it also works well for testing component outputs or content projection, which happens often. Knowing that mount will be called potentially in many places (each test case), it would be nice to offer a straightforward API with minimal boilerplate:

```ts
mount(`<p highlightTypos>{{value}}<p>`, {
  componentProperties: {value: 'something'},
  imports: [HighlightTyposDirective],
});
```

Reply:
Hey guys, I don't want to be annoying here, but I still think that it's wiser to delay the newly introduced API's and prioritise merging first, considering it's quite a significant decision. This could still be revisited at a later stage if there is demand and alignment with the Playwright team. 

CC @pavelfeldman

Reply:
I agree that the priority is to merge this as soon as we can. Indeed, Pavel's point of view on this would help a lot.

Here are the main strategic problems if we do not include the template, `providers` and `imports`:
- Users will have to create way more wrapper components compared to using the template approach. In addition to the extra boilerplate and the indirection that might harm readability, I am afraid that this would harm the adoption. 
Especially that users are now used to template rendering: [Angular Testing Library](https://testing-library.com/docs/angular-testing-library/api#template), [Cypress CT](https://docs.cypress.io/guides/component-testing/angular/api#mount:~:text=%3Capp%2Dstepper%3E%3C/app%2Dstepper%3E)
- To override providers (which will be necessary for most tests), users will have to use hooks. This has the following drawbacks: steep learning curve, boilerplate, indirection, no way to provide generic automatic migrations to future APIs.
- Most users are using standalone components and importing the desired component/directive in the module is not enough. Users must override the component's imports with `TestBed.overrideComponent` which makes the hook harder to implement.
- Stripping the PR from the template/providers/imports APIs  delay 😅

@sand4rt by "newly introduced APIs", you mean compared to other frameworks' support, right?

Reply:
> @sand4rt by "newly introduced APIs", you mean compared to other frameworks' support, right?

I meant the APIs that are newly introduced compared to the target branch. Let's wait a bit, but considering how busy the Playwright team is, it might take a while before we get a response.

The other frameworks share similar features/needs and they don't have dedicated APIs for this. So i do think that if Angular receives dedicated APIs, the other frameworks should have them as well.

Other frameworks currently support similar functionality through the hooksConfig. For Angular, it would probably resemble the code below:

```ts
// playwright/index.ts
beforeMount(async ({ TestBed, hooksConfig }) => {
  TestBed.configureTestingModule({ // <-- something like this could be generated with create-playwright
    imports: hooksConfig?.imports,
    providers: hooksConfig?.providers
  });
});
```

```ts
test('hooks config', async ({ mount }) => {
  const component = await mount(AngularComponent, {
    hooksConfig: {
      imports: [Component],
      providers: [Service],
    }
  });
  await expect(component).toContainText('boop');
});
```

Reply:
> I meant the APIs that are newly introduced compared to the target branch. Let's wait a bit, but considering how busy the Playwright team is, it might take a while before we get a response.

Oh ok! So they are not really newly introduced as the whole Angular support is newly introduced 😄

> The other frameworks share similar features/needs and they don't have dedicated APIs for this. So i do think that if Angular receives dedicated APIs, the other frameworks should have them as well.

Except for Svelte, other template-supporting frameworks (i.e. Vue) can still use JSX.
Now back to Svelte, it would indeed come in handy to allow such a thing by compiling an inline source. Angular template support might inspire such an initiative 😉

> Other frameworks currently support similar functionality through the hooksConfig.

It is not as straightforward when using Angular standalone components.

Indeed, users must override the component's imports and not just update the testbed's module imports which would look something like this:

```ts
beforeMount(async ({TestBed, hooksConfig, component}) => {
  TestBed.overrideComponent(component, hooksConfig.componentOverrides);
});

...

await mount(AngularComponent, {
  hookConfig: {
    componentOverrides: {
      imports: {set: [Component]}
    }
  }
});
```

which raises many questions:
- should playwright CT pass the component to the beforeMount hook?
- what should be the generated boilerplate do? pass the imports that are set? or pass `add`, `remove`, `set` instructions?
- knowing that we can't pass `add`, `remove`, `set` in the same `overrideComponent` call, shouldn't it be an array of overrides?

I am afraid that any hooks based boilerplate or documentation will confuse and discourage users compared to simply pass a template which is pretty common in the Angular community as mentioned before. Especially, that it is not possible to declare a component in the test file (which we could extract by updating the babel transform but I wouldn't recommend it and I don't think that the Playwright team would agree to overspecify the transform with Angular specifics).

Also, note that even Angular's core itself uses templates to make some tests easier to read and maintain.
Here is an interesting example: 
https://github.com/angular/angular/blob/fd544159e300340a25a48e3f40c58795991a4c17/packages/common/test/directives/ng_if_spec.ts#L144

To port these tests to Playwright CT without the template mount, the only alternatives would be to create 14 components or create a component in the `index.ts` or another file and override the template using a hook.
The second option is probably better but it is:
- boilerplat-y, 
- less readable (one has to read the test, the component, the hook's behavior in order to understand)
- more importantly not beginner friendly. Few users know that they can override templates.
- less future-proof: by passing the template to `mount`, we handle the creation of the component in the best way (we might compile it or let JIT compiler do it, we might create a wrapper component and override the template etc...) and we can adapt to future Angular's TestBed changes.

Reply:
@sand4rt do you think you can merge this as is into the main PR so we can bring more people from Playwright and the community to the discussion? or do you really want us to remove the template/providers/imports for merge this?

Reply:
@yjaaidi I still think its better to merge this PR without the template/providers/imports/slots indeed. Then we could merge the other PR with Playwright so people could start trying it out while the template/providers/imports APIs are discussed.

Reply:
ok

Reply:
@yjaaidi did it in https://github.com/yjaaidi/playwright/pull/1, feel free to copy/close the PR if you wanted to do it

Reply:
oh super cool! Thanks. This PR waterfall is memorable 😅

What we finally did is that we updated `@jscutlery/playwright-ct-angular` with all the features and made it align more with other libs (`inputs` => `props` etc...)

By the way, we improved two things there that we didn't backport here:
- more precise transform (we are not transforming function calls in mount args)
- type-safe props & listeners 

but we can add these improvements later to make this ship faster.

Reply:
Just merged! Thanks for the PR 😊.

Where do you live by the way? Once we ship this we should grab a beer with @edbzn and @brandonroberts to celebrate it 🥳

Reply:
Haha yeah we were like 3 levels deep in at some point :) Would be fun! but I think we live quite far from each other. I think @brandonroberts is from the US and im from EU/Netherlands.

Reply:
A celebration is definitely in order 😊

Reply:
We'll meet at a conference some day 😊

Reply:
Hey @timdeschryver, appreciate the feedback! Your insights are definitely valuable here, and i'd love to hear more. Feel free to share additional thoughts on https://github.com/microsoft/playwright/pull/27783.

Also wondering if you have thought about applying the template API to other frameworks such as Svelte and Vue?